File: test/unit/lua/gc_optimized_set_spec.lua
local GCOptimizedSet = require("tstl-gc-optimized-collections.lua.gc_optimized_set").GCOptimizedSet

describe("GCOptimizedSet", function()
    describe("constructor", function()
        local my_set = GCOptimizedSet.new(1, 2, 3)

        it("should have size 3", function()
            assert.is.equal(3, my_set.size)
        end)

        it("should contain all the elements", function()
            assert.is_true(my_set:has(1))
            assert.is_true(my_set:has(2))
            assert.is_true(my_set:has(3))
        end)
    end)

    describe("entries", function()
        local my_set = GCOptimizedSet.new(4, 2, 8)

        it("iterates over the entries", function()
            local set_sum = 0
            for k in my_set:entries() do
                set_sum = set_sum + k
            end
            assert.is.equal(14, set_sum)
        end)
    end)

    describe("add", function()
        local my_set = GCOptimizedSet.new(4, 2, 8)

        my_set:add(10)

        it("adds the element to the set", function()
            assert.is_true(my_set:has(10))
        end)

        it("increases the set size by 1", function()
            assert.is.equal(4, my_set.size)
        end)
    end)

    describe("clear", function()
        local my_set = GCOptimizedSet.new(4, 2, 8)

        my_set:clear()

        it("removes the elements", function()
            assert.is_false(my_set:has(4))
            assert.is_false(my_set:has(2))
            assert.is_false(my_set:has(8))
        end)

        it("sets the size to 0", function()
            assert.is.equal(0, my_set.size)
        end)
    end)

    describe("delete", function()
        local my_set = GCOptimizedSet.new(4, 2, 8)

        my_set:delete(4)

        it("removes the element", function()
            assert.is_false(my_set:has(4))
        end)

        it("decreases the size by 1", function()
            assert.is.equal(2, my_set.size)
        end)
    end)

    describe("has", function()
        local my_set = GCOptimizedSet.new(1, 2, 3)

        describe("the element exists in the set", function()
            it("returns true", function()
                assert.is_true(my_set:has(3))
            end)
        end)

        describe("the element does not exist in the set", function()
            it("returns false", function()
                assert.is_false(my_set:has(5))
            end)
        end)
    end)
end)

File: test/unit/ts/gc_optimized_set_spec.ts
import { GCOptimizedSet } from "../../../tstl-gc-optimized-collections/init";

describe("GCOptimizedSet", () => {
    describe("Empty", () => {
        it("should be empty", () => {
            expect(GCOptimizedSet.Empty.size).toBe(0);
        });
    });

    describe("constructor", () => {
        describe("constructed with no argument", () => {
            const mySet = new GCOptimizedSet();

            it("should be size 0", () => {
                expect(mySet.size).toBe(0);
            });
        });

        describe("constructed with iterable", () => {
            const mySet = new GCOptimizedSet("a", "b");

            it("should have all given elements", () => {
                expect(mySet.has("a")).toBeTruthy();
                expect(mySet.has("b")).toBeTruthy();
            });
        });
    });

    describe("add", () => {
        const mySet = new GCOptimizedSet();
        mySet.add("a");

        it("should have element", () => {
            expect(mySet.has("a")).toBeTruthy();
        });
    });

    describe("clear", () => {
        const mySet = new GCOptimizedSet<number>(0, 1);

        mySet.clear();

        it("should have size 0", () => {
            expect(mySet.size).toBe(0);
        });

        it("should not have element 0", () => {
            expect(mySet.has(0)).toBeFalsy();
        });

        it("should not have element 1", () => {
            expect(mySet.has(1)).toBeFalsy();
        });
    });

    describe("delete", () => {
        const mySet = new GCOptimizedSet("a", "b");
        mySet.delete("a");

        it("should not have deleted element", () => {
            expect(mySet.has("a")).toBeFalsy();
        });
    });

    describe("entries", () => {
        const mySet = new GCOptimizedSet(2, 3, 4);

        describe("one argument callback", () => {
            let count = 0;
            for (const v of mySet.entries()) {
                count += v;
            }

            it("should iterate all entries", () => {
                expect(count).toBe(9);
            });
        });

        describe("two argument callback", () => {
            let count = 0;
            for (const v of mySet.entries()) {
                count += v;
            }

            it("should iterate all entries", () => {
                expect(count).toBe(9);
            });
        });
    });

    describe("has", () => {
        const mySet = new GCOptimizedSet("a", "b");

        describe("set has element", () => {
            it("returns true", () => {
                expect(mySet.has("a")).toBeTruthy();
            });
        });

        describe("set does not have element", () => {
            it("returns false", () => {
                expect(mySet.has("c")).toBeFalsy();
            });
        });
    });

    describe("size", () => {
        describe("set is empty", () => {
            const mySet = new GCOptimizedSet();

            it("returns 0", () => {
                expect(mySet.size).toBe(0);
            });
        });

        describe("set has 1 element added", () => {
            const mySet = new GCOptimizedSet();
            mySet.add(1);

            it("returns 1", () => {
                expect(mySet.size).toBe(1);
            });
        });

        describe("set constructed with 2 elements", () => {
            const mySet = new GCOptimizedSet(1,2);

            it("returns 2", () => {
                expect(mySet.size).toBe(2);
            });
        });

        describe("set constructed with 2 elements and cleared", () => {
            const mySet = new GCOptimizedSet(1,2);
            mySet.clear();

            it("returns 0", () => {
                expect(mySet.size).toBe(0);
            });
        });

        describe("set constructed with 2 elements and 1 is deleted", () => {
            const mySet = new GCOptimizedSet(1,2);
            mySet.delete(1);

            it("returns 1", () => {
                expect(mySet.size).toBe(1);
            });
        });
    });
});
